src/datarobotx/idp/custom_application_source.py

```diff
@@ -0,0 +1,316 @@
+    except KeyError:
+        upload_data: List[Tuple[str, Any]] = []
+        with contextlib.ExitStack() as stack:
+            if folder_path:
+                for root_path, _, file_paths in os.walk(folder_path):
+                    for path in file_paths:
+                        file_path = os.path.join(root_path, path)
+                        file = stack.enter_context(open(file_path, "rb"))
+                        upload_data.append(("file", (os.path.basename(file_path), file)))
+                        upload_data.append(("filePath", os.path.relpath(file_path, folder_path)))
+
+            if runtime_parameter_values_objs:
+                upload_data.append(
+                    (
+                        "runtimeParameterValues",
+                        json.dumps(
+                            [
+                                {camelize(k): v for k, v in param.to_dict().items()}
+                                for param in runtime_parameter_values_objs
+                            ]
+                        ),
+                    )
+                )
+
+            if "replicas" in kwargs:
+                upload_data.append(("replicas", str(kwargs["replicas"])))
+
+            if base_environment_id := kwargs.get("base_environment_id"):
+                upload_data.append(("baseEnvironmentId", base_environment_id))
+
+            if base_environment_version_id := kwargs.get("base_environment_version_id"):
+                upload_data.append(("baseEnvironmentVersionId", base_environment_version_id))
+            upload_data.append(
+                (
+                    "label",
+                    f"{kwargs.get('label', '')}{' - ' if kwargs.get('label') else ''}[{version_token}]",
+                )
+            )
+
+            encoder = MultipartEncoder(fields=upload_data)
+            headers = {"Content-Type": encoder.content_type}
+
+            if previous_custom_application_source_version_id is not None:
+                response = client.request(
+                    method="PATCH",
+                    url=f"customApplicationSources/{custom_application_source_id}/versions/{previous_custom_application_source_version_id}/",
+                    data=encoder,
+                    headers=headers,
+                )
+            else:
+                response = client.request(
+                    method="POST",
+                    url=f"customApplicationSources/{custom_application_source_id}/versions/",
+                    data=encoder,
+                    headers=headers,
+                )
+            new_version = response.json()["id"]
+
+        return str(new_version)
+
+
+# Public functions
+def get_or_create_custom_application_source_version(
+    endpoint: str,
+    token: str,
+    custom_application_source_id: str,
+    runtime_parameter_values: Optional[List[Union[RuntimeParameterValue, Dict[str, str]]]] = None,
+    previous_custom_application_source_version_id: Optional[str] = None,
```

Comment:
this parameter is not present in the rest API (is idiosyncratic to idp) and adds a lot complexity (risk surface for mistakes); i don't think we have this capability for custom model version - do we really need it here?